### PR TITLE
deliveryplanning shortage overage

### PR DESCRIPTION
**Pull request by mrie04**
https://github.com/metasfresh/metasfresh/issues/14492

### CODE REVIEW

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+		inventoryLineRecord.setAD_Org_ID(inventoryRecord.getAD_Org_ID());
+		inventoryLineRecord.setM_Locator_ID(locatorId.getRepoId());
+		inventoryLineRecord.setM_Product_ID(productId.getRepoId());
+		inventoryLineRecord.setM_AttributeSetInstance_ID(asiId.getRepoId());
+		inventoryLineRecord.setHUAggregationType(HUAggregationType.toCodeOrNull(HUAggregationType.SINGLE_HU));
+		inventoryLineRecord.setC_UOM_ID(qtyBooked.getUomId().getRepoId());
+		inventoryLineRecord.setQtyCount(qtyBooked.toBigDecimal().add(p_Qty));
+		inventoryLineRecord.setQtyBook(qtyBooked.toBigDecimal());
+		inventoryLineRecord.setIsCounted(true);
+		inventoryLineRecord.setM_HU_ID(huIds.stream().findFirst().get().getRepoId());
```
Comment (metas-ts) [inline backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:280]: - better use `hus` maybe you don'T even need a dedicated variable for `huIds`
- use `CollectionUtils#singleElement(java.util.Collection<T>)` to get he single item and assert that there is just one

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+				.huForInventoryLineFactory(huForInventoryLineFactory)
+				.huIds(huIds)
+				.build();
+		final List<HuForInventoryLine> hus = shortageAndOverageStrategy.streamHus()
+				.collect(ImmutableList.toImmutableList());
+
+		final Quantity qtyBooked = computeTotalQtyBooked(hus, stockingUOM, UOMConversionContext.of(productId));
+		if(qtyBooked.toBigDecimal().add(p_Qty).signum() < 0)
+		{
+			throw new AdempiereException("Shortage is greater than booked quantity");
```
Comment (metas-ts) [inline backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:266]: `AD_Message` pls

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java (hunk 1 above)
Comment (metas-ts) [reply backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:280]: Later edit: as talked, please check out `DraftInventoryBase` and its existing subclasses, and create a new one. if that doesn't work, let's talk about why it doesn't

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/draftlinescreator/ShortageAndOverageStrategy.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		Check.errorUnless(warehouseId.equals(locatorId.getWarehouseId()),
+							  "The locator's WH needs to be the given WH; warehouseId={}; locatorId={}",
+							  warehouseId, locatorId);
+
+	}
+
+	@Override
+	public Stream<HuForInventoryLine> streamHus()
+	{
+		final IHUQueryBuilder huQueryBuilder = handlingUnitsDAO.createHUQueryBuilder()
```
Comment (metas-ts) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/inventory/draftlinescreator/ShortageAndOverageStrategy.java:100]: why create this query? AFAIU you already have the respecitve IDs right from the start 🤔

**Review by metas-ts — CHANGES_REQUESTED** [review]
TL;DR: please check out `DraftInventoryBase` and its existing subclasses, and create a new one. 
If that doesn't work, let's talk about why it doesn't

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+		final I_M_Inventory inventoryRecord = newInstance(I_M_Inventory.class);
+		inventoryRecord.setAD_Org_ID(receiptInfo.getOrgId().getRepoId());
+		inventoryRecord.setC_DocType_ID(doctypeId.getRepoId());
+		inventoryRecord.setM_Warehouse_ID(inOutRecord.getM_Warehouse_ID());
+		inventoryRecord.setMovementDate(inOutRecord.getMovementDate());
+
+		saveRecord(inventoryRecord);
+		logger.trace("Insert inventory - {}", inventoryRecord);
+
+		createInventoryLine(InventoryId.ofRepoId(inventoryRecord.getM_Inventory_ID()), inOutLines.get(0));
```
Comment (metas-ts) [inline backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:209]: 1st of all: use Collection.singleElement()
2nd of all: you don't need to handle I_M_InOutLine - as i said before, please try to use them as little as possible outside of repos. In thista case you just need the `de.metas.inout.InOutLineId` in order to create your `TableRecordReference`

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	private I_M_Inventory generateInventoryDocument(@NonNull final DocTypeId doctypeId, @NonNull final DeliveryPlanningId deliveryPlanningId)
+	{
+		final DeliveryPlanningReceiptInfo receiptInfo = deliveryPlanningService.getReceiptInfo(deliveryPlanningId);
+		final InOutId inOutId = receiptInfo.getReceiptId();
+		Check.assumeNotNull(inOutId, "InOutId shall be set, because of isReceived() check in preconditions");
+
+		final I_M_InOut inOutRecord = load(inOutId, I_M_InOut.class);
+		final List<I_M_InOutLine> inOutLines = inOutDAO.retrieveLines(inOutRecord);
```
Comment (metas-ts) [inline backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:190]: instead of loading the lines (=> see comment below) you can invoke IInOutDAO.retrieveLinesForInOutId and just get the IDs - not the whole uneeded record

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+		return inventoryRecord;
+	}
+
+	private void createInventoryLine(@NonNull final InventoryId inventoryId, @NonNull final I_M_InOutLine inOutLineRecord)
+	{
+		final TableRecordReference mInOutLineTableRecordReference = TableRecordReference.of(inOutLineRecord);
+		final Set<TableRecordReference> mInOutLineTableRecordReferenceSet = Collections.singleton(mInOutLineTableRecordReference);
+		final ImmutableSetMultimap<TableRecordReference, HuId> huIdsMultimap = huAssignmentDAO.retrieveHUsByRecordRefs(mInOutLineTableRecordReferenceSet);
+		final Collection<HuId> huIds = huIdsMultimap.get(mInOutLineTableRecordReference);
+		final I_M_HU huRecord = load(CollectionUtils.singleElement(huIds), I_M_HU.class);
```
Comment (metas-ts) [inline backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:220]: you don't need the HUrecord in here, so better pass the HuId to the strategy, and load the record only when you really, really need it

**Review by metas-ts — APPROVED** [review]
I do have some requests, but i think you can refactore them in in a 2nd step.
Please make sure that step comes soon.

File: backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java (hunk 6 above)
Comment (mrie04) [reply backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:220]: i thought it would be better to load it once when the strategy is created instead of every time when `de.metas.handlingunits.inventory.draftlinescreator.ShortageAndOverageStrategy#streamHus()` needs it for `huForInventoryLineFactory.ofHURecordWithQtyAdjustment`

Comment (metas-ts) [reply backend/de.metas.deliveryplanning.base/src/main/java/de/metas/deliveryplanning/process/M_Delivery_Planning_GenerateShortageOverage.java:220]: before i forget: if you load it, please use `IHandlingUnitsDAO#getById`.
